### PR TITLE
Add typ package

### PR DESCRIPTION
**Pull request by danking**
I'm curious what people think about this

Did this on one of my flights. Moved all the types to a separate package (so that you can depend on hail types without depending on the AST/IR). When doing this with IntelliJ, I had to move each class to a separate file. This doesn't seem terrible actually because `Type.scala` is rather large and makes IntelliJ angry. On the other hand, we can't seal `Type` if its subclasses are not in the same file.

cc:  @tpoterba @cseed

### CODE REVIEW

**Comment by tpoterba** [comment]
I'm fine with this. IntelliJ struggles hard enough.

The abbrev "typ" makes me irrationally angry for some reason, though. Possibly "types"?

**Comment by danking** [comment]
`is.hail.types` is fine with me.

File: src/main/scala/is/hail/expr/typ/TBinaryOptional.scala [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package is.hail.expr.typ
+
+case object TBinaryOptional extends TBinary(false)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/typ/TBinaryOptional.scala]: Seems slightly excessive to put these in their own file.

**Comment by cseed** [comment]
`types`  is fine with me, too.

**Comment by danking** [comment]
@cseed @tpoterba addressed.

**Pull request merged**